src/test/java/com/puppycrawl/tools/checkstyle/xpath/XpathMapperTest.java

```diff
@@ -45,6 +46,28 @@ protected String getPackageLocation() {
         return "com/puppycrawl/tools/checkstyle/xpath/xpathmapper";
     }
 
+    @Test
+    public void testNodeOrdering() throws Exception {
+        final String xpath = "//METHOD_DEF[./IDENT[@text='callSomeMethod']]/SLIST/*";
```

Comment:
This test fails for the xpath "//METHOD_DEF/SLIST/*"

Reply:
Yes, you are right, I'll revert back to the recursion routine.

Reply:
How about more depth?
```
CLASS_DEF0
  CLASS_DEF1
     METHOD_DEF1-1
         VARIABLE_DEF1-1-1
         VARIABLE_DEF1-1-2
     METHOD_DEF1-2
         VARIABLE_DEF1-2-1
         VARIABLE_DEF1-2-2
  CLASS_DEF2
     METHOD_DEF2-1
         VARIABLE_DEF2-1-1
         VARIABLE_DEF2-1-2
     METHOD_DEF2-2
         VARIABLE_DEF2-2-1
         VARIABLE_DEF2-2-2
```
Now all `VARIABLE_DEF`s are on 4th depth, their parents on 3rd depth.
We can not just compare order of `METHOD_DEF`s, we should go up till a common parent node.

Reply:
@pbludov When the depth is same I have computed their common ancestor's `indexAmongSiblings`. Since the children to be compared are at the same level each call will have a time complexity of `O(heightOfTree)` i.e., logarithmic time and hence won't be much heavy.

Reply:
@pbludov Please review.

Reply:
> How about more depth?
> 
> ```
> CLASS_DEF0
>   CLASS_DEF1
>      METHOD_DEF1-1
>          VARIABLE_DEF1-1-1
>          VARIABLE_DEF1-1-2
>      METHOD_DEF1-2
>          VARIABLE_DEF1-2-1
>          VARIABLE_DEF1-2-2
>   CLASS_DEF2
>      METHOD_DEF2-1
>          VARIABLE_DEF2-1-1
>          VARIABLE_DEF2-1-2
>      METHOD_DEF2-2
>          VARIABLE_DEF2-2-1
>          VARIABLE_DEF2-2-2
> ```
> 
> Now all `VARIABLE_DEF`s are on 4th depth, their parents on 3rd depth.
> We can not just compare order of `METHOD_DEF`s, we should go up till a common parent node.

Yes, I realized this case and deleted my comment just after a while and implemented this way.